### PR TITLE
feat(migration): add modular payment methods migration API support 

### PR DESCRIPTION
**Pull request by GauravRawat369**
## Type of Change

- [ ] Bugfix
- [x] New feature
- [ ] Enhancement
- [ ] Refactoring
- [ ] Dependency updates
- [ ] Documentation
- [ ] CI/CD

## Description

**This pr adds migration api which fixes generic locker data
fix:**
1. Fingerprint problem for bank_debit pm in v1
2. Fix entity_id (v1 its combination of mid+cid, v2 its cid but should be mid)(reminder only for generic locker, legacy still have mid+cid)
3. Fix key for fingerprint

**From v1 what all goes to the generic locker**
1. Bank_debit payment method
2. wallet payment_method

**So what all this api will fix** 
1. V1 pm records for bank_debit and wallets records
2. V2 pm records(cards only)

**How will locker look like**
1. Locker will have new records with new entity_id but same vault_id
2. fingerprint will have new records with same fingerprint but diff hash

### Additional Changes

- [ ] This PR modifies the API contract
- [ ] This PR modifies the database schema
- [ ] This PR modifies application configuration/environment variables

## How did you test it?

Payment in v1
```
curl --location 'http://localhost:8080/payments' \
--header 'Content-Type: application/json' \
--header 'Authorization;' \
--header 'api-key: dev_0wRu0M2vQCwtY83hIfGB0q6xNUyCPunoCONKHoIX9fw7yc75ThfMk11QwrEbz6fg' \
--data-raw '
{
    "amount": 6540,
    "currency": "USD",
    "amount_to_capture": 6540,
    "confirm": true,
    "profile_id": "pro_Ocwc2kGFiSHR4zXOpimw",
    "capture_method": "automatic",
    "capture_on": "2022-09-10T10:11:12Z",
    "authentication_type": "no_three_ds",
    "setup_future_usage":"off_session", 
    
    "customer_id": "0a_cus_019eb8032fd47b519e34c38b6e0afcf6",

    "phone_country_code": "+1",

    "description": "Its my first payment request",
    "return_url": "https://google.com",
    "payment_method": "bank_debit",
    "payment_method_type": "ach",
    "payment_method_data": {

        "bank_debit": {
            "ach_bank_debit": {
                "account_number": "1234567890",
                "routing_number": "011000138",
                
                "billing_default": null,
                "bank_account_holder_name": "gaurav"
            }
        },
        "billing": {
            "address": {
                "line1": "1467",
                "line2": "Harrison Street",
                "line3": "Harrison Street",
                "city": "San Fransico",
                "state": "California",
                "zip": "94122",
                "country": "US",
                "first_name": "joseph",
                "last_name": "Doe"
            },
            "phone": {
                "number": "8056594427",
                "country_code": "+91"
            },
            "email": "guest@example.com"
        }
    },
    
    "billing": {
        "address": {
            "line1": "1467",
            "line2": "Harrison Street",
            "line3": "Harrison Street",
            "city": "San Fransico",
            "state": "California",
            "zip": "94122",
            "country": "US",
            "first_name": "joseph",
            "last_name": "Doe"
        },
        "phone": {
            "number": "8056594427",
            "country_code": "+91"
        },
        "email": "guest@example.com"
    },
    "shipping": {
        "address": {
            "line1": "1467",
            "line2": "Harrison Street",
            "line3": "Harrison Street",
            "city": "San Fransico",
            "state": "California",
            "zip": "94122",
            "country": "US",
            "first_name": "joseph",
            "last_name": "Doe"
        },
        "phone": {
            "number": "8056594427",
            "country_code": "+91"
        },
        "email": "guest@example.com"
    },
    "statement_descriptor_name": "joseph",
    "statement_descriptor_suffix": "JS",
    "order_details": [
        {
            "product_name": "Apple iphone 15",
            "quantity": 1,
            "amount": 6540,
            "account_name": "transaction_processing"
        }
    ],
    "metadata": {
        "udf1": "value1",
        "new_customer": "true",
        "login_date": "2019-09-10T10:11:12Z"
    },
    "browser_info": {
        "user_agent": "Mozilla\/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit\/537.36 (KHTML, like Gecko) Chrome\/70.0.3538.110 Safari\/537.36",
        "accept_header": "text\/html,application\/xhtml+xml,application\/xml;q=0.9,image\/webp,image\/apng,*\/*;q=0.8",
        "language": "nl-NL",
        "color_depth": 24,
        "screen_height": 723,
        "screen_width": 1536,
        "time_zone": 0,
        "java_enabled": true,
        "java_script_enabled": true,
        "ip_address": "128.0.0.1"
    },
    "customer_acceptance": {
        "acceptance_type": "offline",
        "accepted_at": "1963-05-03T04:07:52.723Z",
        "online": {
            "ip_address": "125.0.0.1",
            "user_agent": "amet irure esse"
        }
    },
    "payment_type": "normal", 
    "merchant_order_reference_id": "test_ord",
    "session_expiry": 900  
}'
```
res
```
{
    "payment_id": "pay_GMBhyp6wI3tqwEsgzmxc",
    "merchant_id": "merchant_1781203623",
    "status": "succeeded",
    "amount": 6540,
    "net_amount": 6540,
    "shipping_cost": null,
    "amount_capturable": 0,
    "amount_received": 6540,
    "processor_merchant_id": "merchant_1781203623",
    "initiator": null,
    "sdk_authorization": "cHJvZmlsZV9pZD1wcm9fT2N3YzJrR0ZpU0hSNHpYT3BpbXcscHVibGlzaGFibGVfa2V5PXBrX2Rldl83YTRlNTIxMzY3NzY0YzEyYjViNGU5NGNlM2M3ODc3YixjbGllbnRfc2VjcmV0PXBheV9HTUJoeXA2d0kzdHF3RXNnem14Y19zZWNyZXRfVEhTSFJsNnZ1WHc0S2RGdWlBSEUsY3VzdG9tZXJfaWQ9MGFfY3VzXzAxOWViODAzMmZkNDdiNTE5ZTM0YzM4YjZlMGFmY2Y2LGNsaWVudF9zZXNzaW9uX2lkPWNsaWVudF9zZXNzX3lZRnN5ZGM5dkRpRlplaEN3TEFsLHBheW1lbnRfaWQ9cGF5X0dNQmh5cDZ3STN0cXdFc2d6bXhj",
    "connector": "adyen",
    "state_metadata": null,
    "client_secret": "pay_GMBhyp6wI3tqwEsgzmxc_secret_THSHRl6vuXw4KdFuiAHE",
    "created": "2026-06-11T19:19:40.325Z",
    "modified_at": "2026-06-11T19:19:42.306Z",
    "connector_customer_id": null,
    "currency": "USD",
    "customer_id": "0a_cus_019eb8032fd47b519e34c38b6e0afcf6",
    "customer": {
        "id": "0a_cus_019eb8032fd47b519e34c38b6e0afcf6",
        "name": "John Doe",
        "email": "guest@example.com",
        "phone": "999999999",
        "phone_country_code": "+1",
        "customer_document_details": null
    },
    "description": "Its my first payment request",
    "refunds": null,
    "disputes": null,
    "mandate_id": null,
    "mandate_data": null,
    "setup_future_usage": "off_session",
    "off_session": null,
    "capture_on": null,
    "capture_method": "automatic",
    "payment_method": "bank_debit",
    "payment_method_data": {
        "bank_debit": {
            "ach": {
                "account_number": "******7890",
                "routing_number": "011***138",
                "bank_account_holder_name": "gaurav",
                "bank_name": null,
                "bank_type": null,
                "bank_holder_type": null
            }
        },
        "billing": {
            "address": {
                "city": "San Fransico",
                "country": "US",
                "line1": "1467",
                "line2": "Harrison Street",
                "line3": "Harrison Street",
                "zip": "94122",
                "state": "California",
                "first_name": "joseph",
                "last_name": "Doe",
                "origin_zip": null
            },
            "phone": {
                "number": "8056594427",
                "country_code": "+91"
            },
            "email": "guest@example.com"
        }
    },
    "payment_token": null,
    "shipping": {
        "address": {
            "city": "San Fransico",
            "country": "US",
            "line1": "1467",
            "line2": "Harrison Street",
            "line3": "Harrison Street",
            "zip": "94122",
            "state": "California",
            "first_name": "joseph",
            "last_name": "Doe",
            "origin_zip": null
        },
        "phone": {
            "number": "8056594427",
            "country_code": "+91"
        },
        "email": "guest@example.com"
    },
    "billing": {
        "address": {
            "city": "San Fransico",
            "country": "US",
            "line1": "1467",
            "line2": "Harrison Street",
            "line3": "Harrison Street",
            "zip": "94122",
            "state": "California",
            "first_name": "joseph",
            "last_name": "Doe",
            "origin_zip": null
        },
        "phone": {
            "number": "8056594427",
            "country_code": "+91"
        },
        "email": "guest@example.com"
    },
    "order_details": [
        {
            "sku": null,
            "upc": null,
            "brand": null,
            "amount": 6540,
            "category": null,
            "quantity": 1,
            "tax_rate": null,
            "product_id": null,
            "description": null,
            "product_name": "Apple iphone 15",
            "product_type": null,
            "sub_category": null,
            "total_amount": null,
            "commodity_code": null,
            "unit_of_measure": null,
            "product_img_link": null,
            "product_tax_code": null,
            "total_tax_amount": null,
            "requires_shipping": null,
            "unit_discount_amount": null
        }
    ],
    "email": "guest@example.com",
    "name": "John Doe",
    "phone": "999999999",
    "return_url": "https://google.com/",
    "authentication_type": "no_three_ds",
    "statement_descriptor_name": "joseph",
    "statement_descriptor_suffix": "JS",
    "next_action": null,
    "cancellation_reason": null,
    "error_code": null,
    "error_message": null,
    "unified_code": null,
    "unified_message": null,
    "error_details": null,
    "payment_experience": null,
    "payment_method_type": "ach",
    "connector_label": null,
    "business_country": null,
    "business_label": "default",
    "business_sub_label": null,
    "allowed_payment_method_types": null,
    "manual_retry_allowed": null,
    "connector_transaction_id": "V2FLNLZGD5B3ZMV5",
    "frm_message": null,
    "metadata": {
        "udf1": "value1",
        "login_date": "2019-09-10T10:11:12Z",
        "new_customer": "true"
    },
    "connector_metadata": null,
    "connector_response_metadata": null,
    "feature_metadata": {
        "redirect_response": null,
        "search_tags": null,
        "apple_pay_recurring_details": null,
        "pix_additional_details": null,
        "boleto_additional_details": null,
        "pix_automatico_additional_details": null,
        "finix_additional_details": null
    },
    "reference_id": "pay_GMBhyp6wI3tqwEsgzmxc_1",
    "payment_link": null,
    "profile_id": "pro_Ocwc2kGFiSHR4zXOpimw",
    "surcharge_details": null,
    "attempt_count": 1,
    "merchant_decision": null,
    "merchant_connector_id": "mca_wA26462J4LpNbKdWXUXj",
    "incremental_authorization_allowed": false,
    "authorization_count": null,
    "incremental_authorizations": null,
    "external_authentication_details": null,
    "external_3ds_authentication_attempted": false,
    "expires_on": "2026-06-11T19:34:40.325Z",
    "fingerprint": null,
    "browser_info": {
        "language": "nl-NL",
        "time_zone": 0,
        "ip_address": "128.0.0.1",
        "user_agent": "Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/70.0.3538.110 Safari/537.36",
        "color_depth": 24,
        "java_enabled": true,
        "screen_width": 1536,
        "accept_header": "text/html,application/xhtml+xml,application/xml;q=0.9,image/webp,image/apng,*/*;q=0.8",
        "screen_height": 723,
        "java_script_enabled": true
    },
    "payment_channel": null,
    "payment_method_id": "pm_Ze8j4AotoKBl3jValtq3",
    "network_transaction_id": null,
    "network_transaction_link_id": null,
    "payment_method_status": "active",
    "updated": "2026-06-11T19:19:42.306Z",
    "split_payments": null,
    "frm_metadata": null,
    "extended_authorization_applied": null,
    "extended_authorization_last_applied_at": null,
    "request_extended_authorization": null,
    "capture_before": null,
    "merchant_order_reference_id": "test_ord",
    "order_tax_amount": null,
    "connector_mandate_id": "RX4NRC2SQDPRV4W5",
    "card_discovery": null,
    "force_3ds_challenge": false,
    "force_3ds_challenge_trigger": false,
    "issuer_error_code": null,
    "issuer_error_message": null,
    "is_iframe_redirection_enabled": null,
    "whole_connector_response": null,
    "enable_partial_authorization": null,
    "enable_overcapture": null,
    "is_overcapture_enabled": null,
    "network_details": null,
    "is_stored_credential": null,
    "mit_category": null,
    "billing_descriptor": null,
    "tokenization": null,
    "partner_merchant_identifier_details": null,
    "payment_method_tokenization_details": {
        "payment_method_id": "pm_Ze8j4AotoKBl3jValtq3",
        "payment_method_status": "active",
        "psp_tokenization": false,
        "network_tokenization": false,
        "network_transaction_id": null,
        "network_transaction_link_id": null,
        "is_eligible_for_mit_payment": false
    },
    "installment_options": null,
    "installment_data": null,
    "sender_payment_instrument_id": null
}
```

payment method is pm_Ze8j4AotoKBl3jValtq3
Hit migration api
```
curl --location 'http://localhost:8080/payment_methods/modular/migrate-batch' \
--header 'api-key: test_admin' \
--header 'Accept: application/json' \
--form 'merchant_id="merchant_1781203623"' \
--form 'file=@"/Users/gaurav.rawat/repos/hyperswitch/filename.csv"'
```
res
```
{
    "successfully_migrated": [
        {
            "payment_method_id": "pm_Ze8j4AotoKBl3jValtq3"
        }
    ],
    "failed_migrations": []
}
```

image
Same fingerprint with different hash 1: complete PaymentMethodVaultingData 2. FingerprintData
image

## Checklist

- [ ] I formatted the code `cargo +nightly fmt --all`
- [ ] I addressed lints thrown by `cargo clippy`
- [ ] I reviewed the submitted code
- [ ] I added unit tests for my changes where possible

### CODE REVIEW

**Comment by XyneSpaces** [comment]
## Review Summary

**Classification:** Feature — Modular Payment Method Migration API  
**Risk Level:** Medium (new API contract, event tracking, core flow changes)  
**Files Changed:** Multiple files across api_models, router, diesel_models

---

### Baseline Verification

| Check | Status | Notes |
|-------|--------|-------|
| B1 No unwrap/expect | ✅ | Clean error propagation |
| B2 No unsafe | ✅ | No unsafe blocks |
| B3 error_stack | ✅ | Proper change_context usage |
| B4 Secret<T> | N/A | No new sensitive fields |
| B5 #[instrument] | ⚠️ | See finding below |
| B6 Feature gates | N/A | |
| B7 Newtypes | ✅ | Uses existing type wrappers |

---

### Findings by Severity

#### :warning: Missing #[instrument] on new async core functions

New async functions in `crates/router/src/core/payment_methods.rs` (and related files) handling migration operations should carry `#[instrument(skip_all, fields(...))]` for observability in production.

**Fix:** Add tracing instrumentation:
```rust
#[instrument(skip_all, fields(merchant_id = %merchant_id, migration_id = %migration_id))]
pub async fn migrate_payment_methods_modular(...) { ... }
```

---

#### :warning: Missing test coverage

New feature adding modular payment method migration API — no test files are modified in the PR. At minimum, unit tests for the response struct and API contract validation should be included.

**Fix:** Add tests in `crates/router/tests/` or `cypress-tests/` covering:
- Successful migration response serialization
- Failed migration error handling
- Edge cases (empty input, partial failures)

---

#### :bulb: PR description completeness

The PR body is essentially empty. For a new feature, please provide:
1. **What** — Brief description of the modular migration API
2. **Why** — Problem it solves vs existing migration flow
3. **API contract** — Request/response format, endpoint path
4. **Database changes** — Even if none, explicitly state "No schema changes"
5. **Testing** — How this was validated

---

### Positive Notes

- `ModularPaymentMethodMigrationResponse` struct is well-designed with clear separation of successful/failed migrations  
- `ApiEventMetric` impl properly added for observability  
- Response type uses appropriate serde derives for API consistency

---

**Verdict:** :warning: **Approve with suggestions** — Add instrument attributes and tests before merge. Address PR description completeness.

**Comment by XyneSpaces** [comment]
## Review Summary

**Classification:** Core — Modular Payment Method Migration API
**Risk Level:** Medium-High (new migration flow, vault operations, CSV parsing)
**Files Changed:** 19 files across api_models, domain_models, payment_methods, router
**Scenarios:** `core-flow-framework` + `payment-methods`

---

### Baseline Verification (B1–B7)

| Check | Status | Notes |
|-------|--------|-------|
| B1 File coverage | ✅ | All 19 files reviewed |
| B2 Code integrity | ⚠️ | See C1 finding on CSV parsing error handling |
| B3 Code correctness | ✅ | Proper error_stack usage with `.change_context()` and `.attach_printable()` |
| B4 Companion sync | ✅ | Controller trait methods properly declared in `payment_methods` crate |
| B5 Comment scope | ✅ | Code-only review |
| B6 No dead additions | ✅ | All new code has consumers |
| B7 No scope creep | ✅ | Focused on modular migration feature |

---

### Core Team Checklist Findings

#### Tier 1 — Blocking Issues

**:rotating_light: C1. CSV parsing error handling loses original error context**

`parse_csv_new` uses `csv::Reader::deserialize()` which can fail with detailed CSV errors. The `?` operator propagates these, but the caller `get_payment_method_ids` maps all errors to a generic `PreconditionFailed` without the original CSV error details.

**File:** `crates/payment_methods/src/core/migration.rs:388-394`

**Fix:** Preserve the original error context using `.change_context()` and `.attach_printable()` instead of `.map_err()`.

---

**:rotating_light: C2. Sensitive fingerprint_id passed as plain String**

`get_fingerprint_id_from_vault` and the vault call accept `fingerprint_id: String` instead of `Secret<String>`. Fingerprint IDs can be used to correlate payment methods across merchants and should be treated as PII.

**File:** `crates/payment_methods/src/controller.rs:326-331`, `crates/router/src/core/payment_methods/cards.rs:1495-1501`

**Fix:** Wrap fingerprint_id in `Secret<String>`:
```rust
async fn get_fingerprint_id_from_vault(
    &self,
    entity_id: &V1VaultEntityId,
    fingerprint_data: &FingerprintData,
    fingerprint_id: Secret<String>,
) -> errors::VaultResult<Secret<String>>;
```

---

**:warning: C3. Partial migration failure semantics unclear**

The `modular_migrate_payment_methods` function processes payment methods in a loop and separates successes/failures. Failed migrations return `200 OK` with the `failed_migrations` list. However, the individual failures are logged but not aggregated with error details in the response.

**File:** `crates/payment_methods/src/core/migration.rs:70-93`

Consider enriching the response with error details for each failed migration or adding a "skipped" category.

---

#### Tier 2 — Medium Signal

**:warning: C4. Migration skips non-bank_debit/non-wallet without clear signal**

The `migrate_single_payment_method` function returns an error for non-bank_debit/non-wallet payment methods. This causes them to appear in `failed_migrations` rather than being explicitly skipped.

**File:** `crates/payment_methods/src/core/migration.rs:132-143`

---

**:bulb: C5. `#[cfg(feature = "v1")]` functions in vault.rs lose V2 compatibility**

The removed `#[cfg(feature = "v2")]` guards from `to_fingerprint_data` and `get_bank_debit_fingerprint_data` now expose these functions to V1 builds. Ensure this is intentional for backward compatibility.

**File:** `crates/hyperswitch_domain_models/src/vault.rs`

---

**:mag: C6. No tests for the new migration flow**

The PR introduces a significant new flow with no unit tests or integration tests. Consider adding tests for CSV parsing, vault operations, and migration scenarios.

---

### Positive Notes

- Clean separation of concerns between `migration.rs` and controller trait
- Proper use of `V1VaultEntityId` for vault operations
- Good error propagation with `change_context` and `attach_printable`
- Thread-safe migration loop with individual failure isolation
- Feature-flagged as `v1` for controlled rollout

---

**Verdict:** :warning: **Request Changes** — Address C1 (CSV error context) and C2 (Secret<String> for fingerprint_id) before merge.

**Review by XyneSpaces — COMMENTED** [review]
## Review Summary

**Classification:** Core Feature — Modular Payment Methods Migration API  
**Risk Level:** Medium (new migration flow touching vault operations, payment method data)  
**Files Changed:** 16 files across core, routes, API models, and payment_methods crate  

---

### Baseline Verification (B1–B7)

**B1. File coverage** — All 16 files reviewed  
**B2. Code integrity** — No unwrap/expect in non-test code; clean error propagation  
**B3. Code correctness** — Migration logic follows existing vault patterns  
**B4. Companion sync** — New route registered in app.rs, handler in routes/payment_methods.rs  
**B5. Comment scope** — Code-only review  
**B6. No YAGNI** — All additions have consumers  
**B7. No scope creep** — Focused migration feature  

---

### Core Team Checklist

#### Tier 1 — Blocking Issues

**:rotating_light: Missing feature flag gating for new migration API**

The new `/payment_methods/modular/migrate-batch` endpoint in `routes/app.rs` and the migration logic in `crates/payment_methods/src/core/migration.rs` are not behind a feature flag. Large new features should be gated at the earliest dispatch point to allow gradual rollout.

**Fix**: Add `#[cfg(feature = "modular_migration")]` or similar feature gate on:
- The route registration in `app.rs`
- The public migration functions
- The controller trait methods

---

#### Tier 2 — Medium Signal

**:warning: CSV parsing lacks validation on payment method ID format**

The `parse_csv_new()` function in `migration.rs` parses payment method IDs from CSV but doesn't validate the ID format before processing. Invalid IDs will fail downstream during vault retrieval with less clear error messages.

**Suggestion**: Add validation that IDs match the expected `pay_` prefix format before enqueuing for migration.

---

#### Tier 3 — Nits

**:bulb: `ModularPaymentMethodMigrationResponse` field naming**

The response struct uses `failed_migrations` (plural noun) but `successfully_migrated` (past participle). Consider aligning naming:
- Option A: `successful_migrations` + `failed_migrations`
- Option B: `successfully_migrated` + `failed_to_migrate`

---

### Positive Notes

- Clean separation between controller trait definition and implementation  
- Uses existing vault infrastructure (`call_to_vault`, `AddVaultRequestNew`)  
- Proper event instrumentation with `ApiEventMetric` impl  
- Flow variant added for observability (`ModularPaymentMethodsMigrate`)  
- Backward compatibility preserved for existing migration flows  

---

**Verdict:** :warning: **Request Changes** — Add feature flag gating for the new migration API before merge.

**Review by XyneSpaces — COMMENTED** [review]
## Review Summary

**Classification:** Core Feature — Modular Payment Methods Migration API  
**Risk Level:** Medium-High (payment method migration, vault operations, new API endpoint)  
**Files Changed:** 19 files (+783/-53 lines)  
**Scope:** Payment methods, vault operations, migration framework  

---

### Baseline Verification (B1–B10)

| Check | Status | Notes |
|-------|--------|-------|
| B1 No unwrap/expect | ✅ | Clean error propagation via `?` operator |
| B2 No unsafe without SAFETY | ✅ | No unsafe blocks introduced |
| B3 error_stack preservation | ⚠️ | See finding on error logging in migration loop |
| B4 Secret<T> usage | ✅ | Vault operations use proper encryption |
| B5 #[instrument] | ⚠️ | New async functions missing instrumentation |
| B6 Feature gates | ✅ | `#[cfg(feature = "v1")]` used correctly |
| B7 Newtype wrappers | ✅ | Existing ID types reused |
| B8 ForeignFrom/FforeignTryFrom | ✅ | Type conversions follow patterns |
| B9 Trait objects | N/A | Not applicable to this change |
| B10 Exhaustive matching | ✅ | Payment method type matching is explicit |

---

### Core Team Checklist (Section 4)

#### Tier 1 — Blocking Issues

**:warning: Silent error swallowing in migration batch processing**

`crates/payment_methods/src/core/migration.rs:78-88`

```rust
Err(err) => {
    router_env::logger::error!("Failed to migrate payment method {}: {:?}", pm_id, err);
    failed_migrations.push(pm_id_record);
}
```

Errors are logged but the response only returns success/failed IDs without error details. Callers cannot distinguish between:
- Vault retrieval failures
- Validation failures  
- Payment method not found
- Unsupported payment method types

**Fix:** Include error context in the response:
```rust
pub struct ModularPaymentMethodMigrationResponse {
    pub successfully_migrated: Vec<PaymentMethodId>,
    pub failed_migrations: Vec<FailedMigration>,  // Include error details
}

pub struct FailedMigration {
    pub payment_method_id: PaymentMethodId,
    pub error_code: String,
    pub error_message: String,
}
```

---

**:warning: Missing `#[instrument]` on new async entry points**

New async functions in `migration.rs` lack tracing instrumentation:
- `modular_migrate_payment_methods`
- `migrate_single_payment_method`
- `fetch_payment_method_record`

**Fix:** Add instrumentation for observability:
```rust
#[instrument(skip_all, fields(merchant_id = %merchant_id, payment_method_count = payment_method_ids.len()))]
pub async fn modular_migrate_payment_methods(...)
```

---

**:warning: Incomplete PR description**

The PR description template is unfilled. For a feature of this scope, please provide:
1. What the modular migration API does differently from the existing `/migrate-batch` endpoint
2. Why bank_debit and wallet are the only supported payment method types
3. Migration strategy for existing payment methods
4. Testing approach (unit/integration/manual)

---

#### Tier 2 — Medium Signal

**:bulb: Payment method type restriction logic is brittle**

`crates/payment_methods/src/core/migration.rs:129-142`

The check for bank_debit/wallet uses pattern matching but then returns an error for non-matching types. Consider:
1. Making this a filter rather than an error (skip unsupported types, don't fail)
2. Or document clearly why these specific types are required

Current behavior: If a CSV contains 100 payment methods and 1 is a card, the entire batch fails. Consider partial success semantics.

---

**:bulb: Vault call signature change requires audit trail**

The `call_to_vault` function signature change (adding `additional_headers: Option<String>`) is backward-compatible, but verify:
1. All existing call sites are updated (appears complete in diff)
2. The `X_FINGERPRINT_ID` header is properly documented

---

**:bulb: Consider rate limiting for batch operations**

The endpoint accepts a CSV file with unlimited payment method IDs. Consider:
- Maximum batch size validation
- Rate limiting per merchant
- Progress tracking for large batches

---

#### Tier 3 — Nits

**:mag: Variable naming in controller trait**

`crates/payment_methods/src/controller.rs:311-329`

The trait methods `retrieve_payment_method_from_vault`, `store_payment_method_in_vault`, `get_fingerprint_id_from_vault` could be more concise:
- `retrieve_from_vault`
- `store_in_vault`  
- `get_fingerprint_id`

The vault context is implied by the trait name and parameters.

---

**:mag: `parse_csv_new` function name**

`crates/payment_methods/src/core/migration.rs:422-429`

The name `parse_csv_new` is not descriptive. Consider `parse_payment_method_ids_csv` or similar.

---

### Positive Notes

- Clean separation of concerns between controller, core migration logic, and routes
- Proper use of `V1VaultEntityId` for backward compatibility
- Consistent error propagation with `change_context` and `attach_printable`
- Feature-gated correctly for v1 only
- All existing vault call sites updated consistently

---

### Verification Checklist

- [ ] PR description filled with motivation and context
- [ ] `cargo +nightly fmt --all` run
- [ ] `cargo clippy` clean
- [ ] Unit tests added for `modular_migrate_payment_methods`
- [ ] Integration tests for success/failure scenarios
- [ ] Documentation updated for new endpoint `/modular/migrate-batch`

---

**Verdict:** ⚠️ **Request Changes** — Address blocking issues before merge:
1. Include error details in failed migration response
2. Add `#[instrument]` to new async entry points  
3. Complete PR description with motivation and testing approach

File: crates/payment_methods/src/core/migration.rs [hunk 1]
```diff
@@ -64,6 +66,187 @@ pub async fn migrate_payment_methods(
...
+    // Step 5: Create new record in vault with new entity_id(merchant_id)
+
+    controller
+        .store_payment_method_in_vault(&merchant_id.clone(), &vault_id, &vaulting_data)
+        .await
+        .change_context(errors::ApiErrorResponse::InternalServerError)
+        .attach_printable("Failed to store payment method in vault with new entity_id")?;
+
+    // Step 6 & 7: If bank_debit, handle fingerprint data migration with new customer key
+    if is_bank_debit {
```
Comment (kashif-m) [inline crates/payment_methods/src/core/migration.rs]: Add is_wallet

File: crates/api_models/src/payment_methods.rs [hunk 2]
```diff
@@ -3981,6 +3981,12 @@ pub struct PaymentMethodMigrationResponse {
     pub network_transaction_id_migrated: Option<bool>,
 }
 
+#[derive(Debug, Default, serde::Serialize)]
+pub struct ModularPaymentMethodMigrationResponse {
+    pub successfully_migrated: Vec<PaymentMethodId>,
+    pub failed_migrations: Vec<PaymentMethodId>,
```
Comment (kashif-m) [inline crates/api_models/src/payment_methods.rs]: Push error messages as well

File: crates/router/src/core/payment_methods/vault.rs [hunk 3]
```diff
@@ -1938,6 +1938,7 @@ async fn create_vault_request<R: pm_types::VaultingInterface>(
     payload: Vec<u8>,
     tenant_id: id_type::TenantId,
     write_mode: Option<pm_types::VaultQueryParam>,
+    additional_headers: Option<String>,
```
Comment (kashif-m) [inline crates/router/src/core/payment_methods/vault.rs]:
```suggestion
    additional_headers: Option<HashMap<String, String>>,
```

**Comment by kashif-m** [comment]
Closing this in favor of opening a fresh PR.
